### PR TITLE
added threshold for max size of pair-rdd before similarity computations

### PR DESCRIPTION
**Pull request by mlopatka**
Added thresholding on pair-RDD size to avoid OOM errors on taar_similarity.py

### CODE REVIEW

File: mozetl/taar/taar_similarity.py [hunk 1]
```diff
@@ -26,6 +26,9 @@
 CATEGORICAL_FEATURES = ["geo_city", "locale", "os"]
 CONTINUOUS_FEATURES = \
     ["subsession_length", "bookmark_count", "tab_open_count", "total_uri", "unique_tlds"]
+# Set a constrain on the pairwise comparisons that are allowed
+#  both at an intra- and inter-custer level to avoid OOM errors in KDE.
+KDE_SIZE_THRESH = 1000
```
Comment (Dexterp37) [inline mozetl/taar/taar_similarity.py]: How was this picked? Did you check that this constant helps fixing the OOM situation?

File: mozetl/taar/taar_similarity.py [hunk 2]
```diff
@@ -202,13 +205,18 @@ def safe_get(field, row, default_value):
...
-def generate_non_cartesian_pairs(first_rdd, second_rdd):
+def generate_non_cartesian_pairs(first_rdd, second_rdd, random_seed=None):
     # Add an index to all the elements in each RDD.
     rdd1_with_indices = first_rdd.zipWithIndex().map(lambda p: (p[1], p[0]))
     rdd2_with_indices = second_rdd.zipWithIndex().map(lambda p: (p[1], p[0]))
     # Join the RDDs using the indices as keys, then strip
     # them off before returning an RDD like [<v1, v2>, ...]
-    return rdd1_with_indices.join(rdd2_with_indices).map(lambda p: p[1])
+    rdd_large_out = rdd1_with_indices.join(rdd2_with_indices).map(lambda p: p[1])
+    size_pair_rdd = rdd_large_out.count()
```
Comment (Dexterp37) [inline mozetl/taar/taar_similarity.py]: Counting here doesn't sound great (I know you need that for sampling) :\

**Comment by Dexterp37** [comment]
@mlopatka sounds like this doesn't solve the problem on EMR 5.2.0.

> 17/09/22 09:38:29 WARN ExecutorAllocationManager: Uncaught exception in thread spark-dynamic-executor-allocation
> org.apache.spark.rpc.RpcTimeoutException: Futures timed out after [120 seconds]. This timeout is controlled by spark.rpc.askTimeout
> 	at org.apache.spark.rpc.RpcTimeout.org$apache$spark$rpc$RpcTimeout$$createRpcTimeoutException(RpcTimeout.scala:48)
> 	at org.apache.spark.rpc.RpcTimeout$$anonfun$addMessageIfTimeout$1.applyOrElse(RpcTimeout.scala:63)
> 	at org.apache.spark.rpc.RpcTimeout$$anonfun$addMessageIfTimeout$1.applyOrElse(RpcTimeout.scala:59)
> 	at scala.PartialFunction$OrElse.apply(PartialFunction.scala:167)
> 	at org.apache.spark.rpc.RpcTimeout.awaitResult(RpcTimeout.scala:83)
> 	at org.apache.spark.scheduler.cluster.CoarseGrainedSchedulerBackend.killExecutors(CoarseGrainedSchedulerBackend.scala:591)
> 	at org.apache.spark.SparkContext.killExecutors(SparkContext.scala:1541)
> 	at org.apache.spark.ExecutorAllocationClient$class.killExecutor(ExecutorAllocationClient.scala:65)
> 	at org.apache.spark.SparkContext.killExecutor(SparkContext.scala:1560)
> 	at org.apache.spark.ExecutorAllocationManager.org$apache$spark$ExecutorAllocationManager$$removeExecutor(ExecutorAllocationManager.scala:421)
> 	at org.apache.spark.ExecutorAllocationManager$$anonfun$org$apache$spark$ExecutorAllocationManager$$schedule$1.apply(ExecutorAllocationManager.scala:286)
> 	at org.apache.spark.ExecutorAllocationManager$$anonfun$org$apache$spark$ExecutorAllocationManager$$schedule$1.apply(ExecutorAllocationManager.scala:282)
> 	at scala.collection.mutable.MapLike$$anonfun$retain$2.apply(MapLike.scala:218)
> 	at scala.collection.mutable.MapLike$$anonfun$retain$2.apply(MapLike.scala:217)
> 	at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:733)
> 	at scala.collection.immutable.List.foreach(List.scala:381)
> 	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:732)
> 	at scala.collection.mutable.MapLike$class.retain(MapLike.scala:217)
> 	at scala.collection.mutable.AbstractMap.retain(Map.scala:80)
> 	at org.apache.spark.ExecutorAllocationManager.org$apache$spark$ExecutorAllocationManager$$schedule(ExecutorAllocationManager.scala:282)
> 	at org.apache.spark.ExecutorAllocationManager$$anon$2.run(ExecutorAllocationManager.scala:224)
> 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
> 	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
> 	at java.lang.Thread.run(Thread.java:745)
> Caused by: java.util.concurrent.TimeoutException: Futures timed out after [120 seconds]
> 	at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:219)
> 	at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:223)
> 	at scala.concurrent.Await$$anonfun$result$1.apply(package.scala:190)
> 	at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:53)
> 	at scala.concurrent.Await$.result(package.scala:190)
> 	at org.apache.spark.rpc.RpcTimeout.awaitResult(RpcTimeout.scala:81)
> 	... 23 more
> 17/09/22 09:38:35 WARN TaskMemoryManager: Failed to allocate a page (33554432 bytes), try again.

**Comment by mlopatka** [comment]
Are we sure that the following the RDD assignents inside `get_lr_curves` is overwritign the same (allocated) memory? If this is accidentally generating a new RDD with a unique underlying memory alocation it could be the cause of the OOM error. I do not know enough about the inner workings of spark to assess this.

---- occuring inside a pyhton for loop ----
```
current_cluster_df = (
             features_df
             .where(col("prediction") == cluster_number)
         )
         # Pick the features for users belonging to all the other clusters.
         other_clusters_df = (
             features_df
             .where(col("prediction") != cluster_number)
)
```

**Comment by Dexterp37** [comment]
Closing this in favour of #120

**Pull request closed without merging**